store/postgres/src/chain_store.rs

```diff
@@ -1894,10 +1908,40 @@ impl ChainStoreTrait for ChainStore {
                     .filter(|hash| cached.iter().find(|(ptr, _)| &ptr.hash == *hash).is_none())
                     .cloned()
                     .collect::<Vec<_>>();
-                let stored = self.blocks_from_store(hashes).await?;
-                for block in &stored {
-                    self.recent_blocks_cache.insert_block(block.clone());
-                }
+                // We key this off the entire list of hashes, which means
+                // that concurrent attempts that look up `[h1, h2]` and
+                // `[h1, h3]` will still run two queries and duplicate the
+                // lookup of `h1`. Noticing that the two requests should be
+                // serialized would require a lot more work, and going to
+                // the database for one block hash, `h3`, is not much faster
+                // than looking up `[h1, h3]` though it would require less
+                // IO bandwidth
+                let hash = crypto_stable_hash(&hashes);
+                let this = self.clone();
+                let lookup_fut = async move {
+                    let res = this.blocks_from_store(hashes).await;
+                    BlocksLookupResult(Arc::new(res))
+                };
+                let lookup_herd = self.lookup_herd.cheap_clone();
+                let logger = self.logger.cheap_clone();
+                let (BlocksLookupResult(res), _) =
+                    lookup_herd.cached_query(hash, lookup_fut, &logger).await;
+                // Try to avoid cloning a non-concurrent lookup; it's not
+                // entirely clear whether that will actually avoid a clone
+                // since it depends on a lot of the details of how the
+                // `HerdCache` is implemented
+                let res = Arc::try_unwrap(res).unwrap_or_else(|arc| (*arc).clone());
+                let stored = match res {
+                    Ok(blocks) => {
+                        for block in &blocks {
+                            self.recent_blocks_cache.insert_block(block.clone());
```

Comment:
Why must we clone again here, is `blocks` not owned?

Reply:
We need one copy to stick into the cache, and one copy to return (`blocks()` returns owned values) I thought about sticking that all into an `Arc`, but the use of this in `etherum_adapter` needs an owned value for `serde::from_value`, so one way or another we'll end up cloning